engine.go

```diff
@@ -0,0 +1,390 @@
+	"go.flow.arcalot.io/engine/workflow"
+	"go.flow.arcalot.io/pluginsdk/atp"
+	"go.flow.arcalot.io/pluginsdk/schema"
+)
+
+// WorkflowEngine is responsible for executing workflows and returning their result.
+type WorkflowEngine interface {
+	// RunWorkflow executes a workflow from the passed workflow files as parameters. One of the files must be designated
+	// as a workflow file, which will be parsed from the YAML format. Additional files may be passed so that the
+	// workflow may access them (e.g. a kubeconfig file). The workflow input is passed as a separate file.
+	RunWorkflow(
+		ctx context.Context,
+		input []byte,
+		files map[string][]byte,
+		workflowFileName string,
+	) (
+		outputData any,
+		err error,
+	)
+}
+
+type workflowEngine struct {
+	logger           log.Logger
+	deployerRegistry registry.Registry
+	config           *config.Config
+}
+
+//nolint:funlen,gocognit,gocyclo
+func (w workflowEngine) RunWorkflow(
+	ctx context.Context,
+	input []byte,
+	files map[string][]byte,
+	workflowFileName string,
+) (outputData any, err error) {
+	wf, data, err := w.loadWorkflow(workflowFileName, files)
+	if err != nil {
+		return nil, err
+	}
+
+	inputSchema := wf.Input
+	parsedInput, err := yaml.New().Parse(input)
+	if err != nil {
+		return nil, ErrInvalidInputYAML{
+			err,
+		}
+	}
+	unserializedInput, err := inputSchema.Unserialize(parsedInput.Raw())
+	if err != nil {
+		return nil, ErrInvalidInput{
+			err,
+		}
+	}
+
+	// We lost YAML tags during unserialization, we need to reconstruct them. The output, as well as step input and
+	// deploy will contain YAML nodes after this.
+	output, ok := data.MapKey("output")
+	if !ok {
+		// Map key does not exist.
+		wf.Output = yaml.EmptyNode()
+	} else {
+		wf.Output = output
+	}
+
+	steps, ok := data.MapKey("steps")
+	if !ok {
+		// This should never happen
+		return nil, ErrNoSteps
+	}
+
+	for _, stepID := range steps.MapKeys() {
+		step, _ := steps.MapKey(stepID)
+		if deploy, ok := step.MapKey("deploy"); ok {
+			wf.Steps[stepID].Deploy = deploy
+		}
+		if input, ok := step.MapKey("input"); ok {
+			wf.Steps[stepID].Input = input
+		}
+	}
+
+	stepSchemas, err := w.getStepSchemas(wf)
+	if err != nil {
+		return nil, err
+	}
+
+	w.logger.Infof("Building dependency tree...")
+	// Now we have the YAML tags replaced, let's construct a dependency tree.
+	var depTree dg.DirectedGraph[treeItem]
+	depTree, wf.Steps, wf.Output, err = buildDependencyTree(
+		inputSchema,
+		wf.Steps,
+		stepSchemas,
+		wf.Output.(yaml.Node),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build dependency tree (%w)", err)
+	}
+	if depTree.HasCycles() {
+		return nil, fmt.Errorf("step dependency tree has at least one cycle")
+	}
+	outputNode, err := depTree.GetNodeByID("output")
+	if err != nil {
+		return nil, fmt.Errorf("failed to find output key")
+	}
+	w.logger.Infof("Dependency tree complete.")
+	w.logger.Infof("Dependency tree Mermaid:\n%s", depTree.Mermaid())
+
+	lock := &sync.Mutex{}
```

Comment:
`lock` enables parallel update to `dataStructure` if I understood it correctly. A custom type such as `type ConcurrentDataStructure map[string]any{}` with functions like 
```Go
func (ds ConcurrentDataStructure) Expand(...) { ... }
``` 
could encapsulate the locking and simplify this code. What do you think?